[KAFKA-9644] Non-existent configs in incrementalAlterConfigs APPEND/SUBTRACT

steverod opened the pull request:
Problem
----
The `incrementalAlterConfigs` API supports OpType.APPEND and OpType.SUBTRACT for configuration properties of LIST type. If an APPEND or SUBTRACT OpType is submitted for a config property which currently has no value, then the operation fails with a NullPointerException on the broker side (conveyed as an "unknown server error" to the client).

This is because the alter code does a `getProperty` of the existing configuration value
with no concern as to whether or not the property actually exists.

This change handles the case of existing null properties. 

Testing
-----
This change includes 2 test cases in the unit test that demonstrate the issue for OpType.SUBTRACT and OpType.APPEND. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

steverod committed:
KAFKA-9644: Test case showing SUBTRACT failure.

steverod committed:
KAFKA-9644:  Test case showing NPE in APPEND operation.

steverod committed:
KAFKA-9644: Handle empty properties in incrementalAlterConfigs with OpType.APPEND / OpType.SUBTRACT.

Use a default property value for non-existent properties and strip these out of the
existing property list.

steverod commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -604,14 +604,14 @@ class AdminManager(val config: KafkaConfig,
           if (!listType(alterConfigOp.configEntry().name(), configKeys))
             throw new InvalidRequestException(s"Config value append is not allowed for config key: ${alterConfigOp.configEntry().name()}")
-          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name(), "").split(",").toList.filter(_.trim.length > 0)
```
The filter code is necessary to generate an empty list when there's an empty property.

bob-barrett replied:
Do we need the `trim` call? We should already get a length 0 string back if the property is empty. Seems unlikely that we'd ever have a list value that includes a whitespace-only string as an element, but who knows.

steverod replied:
I'll remove and try.

steverod replied:
Removed the trim and it still passes tests. yay.

steverod commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -604,14 +604,14 @@ class AdminManager(val config: KafkaConfig,
             throw new InvalidRequestException(s"Config value subtract is not allowed for config key: ${alterConfigOp.configEntry().name()}")
-          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name(), "").split(",").toList.filter(_.trim.length > 0)
           val newValueList = oldValueList.diff(alterConfigOp.configEntry().value().split(",").toList)
```
An open question is if we subtract everything, should that remove the property or just leave it as an empty string?

omkreddy replied:
I think, we should remove the config.

steverod replied:
After spending several hours arguing with the test, I realized that a "removal" (delete) reverts the property to the default. So if we have a property whose list value is "foo, bar" and we SUBTRACT "foo, bar" ... is the property now set to "" or to <default>? (Especially if default IS, say, "foo")

omkreddy replied:
I think, we should fallback to broker/topic defaults after removal of any overrides. currently, we should the same for  OpType.DELETE operation. We should also follow the same for OpType. SUBTRACT?

bob-barrett replied:
I think it makes more sense to end up with an empty list if you SUBTRACT each list element. That's how lists typically work. We have DELETE if the desired behavior is to actually remove the config.

omkreddy replied:
My main concern is setting empty list to configs like `cleanup.policy` may lead to unwanted errors. 

cc @rajinisivaram

rajinisivaram replied:
May be worth looking at all the broker and topic configs of type LIST and seeing whether empty list makes sense in most cases. If it doesn't. perhaps throwing an exception may be better than deleting quietly since you may end up with a default value that you didn't intend for the config.

steverod replied:
Thanks @rajinisivaram ! -- a discussion with others led to the realization that empty string is the wrong answer in the general case, I have updated to use the stringified version of the config default.

steverod replied:
I have just pushed a version which stays with the "subtract from the list even if the list goes to empty", and will momentarily push a version which does the "config removal/revert to default when subtract yields an empty list". I defer to the Kafka committers on this.

omkreddy replied:
I think it makes sense to apply append/subtract on the config default
value rather than "", if the config is previously not set. 

The only issue with config removal/revert to default is we can not set empty list to config value.

steverod replied:
We could always force users to use the `SET` operation in this case, rather than `SUBTRACT`.
However, on thinking about it, the current code will let you SUBTRACT to an empty list. So in the interest of keeping this a bug fix (not breaking existing code/changing behavior, just preventing the NPE), maybe we should not include my last pushed commit and allow SUBTRACT to generate an empty list. This is rope, but it is existing rope.

omkreddy replied:
makes sense. we can revert last commit and handle separately in another JIRA.

Can we please update the javadoc with current behaviour of `OpType.SET`, 
`OpType.DELETE`, `OpType.APPEND` and `OpType.SUBTRACT` operations?
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/Admin.java#L394

steverod replied:
FYI, validated in the Javadoc for AlterConfigOp.OpType --

![image](https://user-images.githubusercontent.com/42848865/76537037-685c9480-643a-11ea-842b-e0376580bf61.png)

omkreddy commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -1653,9 +1653,11 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
       new AlterConfigOp(new ConfigEntry(LogConfig.MinCleanableDirtyRatioProp, "0.9"), AlterConfigOp.OpType.SET),
-      new AlterConfigOp(new ConfigEntry(LogConfig.CompressionTypeProp, "lz4"), AlterConfigOp.OpType.SET)
+      new AlterConfigOp(new ConfigEntry(LogConfig.CompressionTypeProp, "lz4"), AlterConfigOp.OpType.SET),
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.APPEND)
```
Can we add a assert statement below to verify that config got added?

steverod replied:
Done.

omkreddy commented on core/src/test/scala/integration/kafka/api/PlaintextAdminIntegrationTest.scala:
```diff
@@ -1681,7 +1683,8 @@ class PlaintextAdminIntegrationTest extends BaseAdminIntegrationTest {
     topic1AlterConfigs = Seq(
-      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.SUBTRACT)
+      new AlterConfigOp(new ConfigEntry(LogConfig.CleanupPolicyProp, LogConfig.Compact), AlterConfigOp.OpType.SUBTRACT),
+      new AlterConfigOp(new ConfigEntry(LogConfig.LeaderReplicationThrottledReplicasProp, "0"), AlterConfigOp.OpType.SUBTRACT)
```
Can we add a assert statement to verify config doesn't exist before and after the alter call?

steverod replied:
Done.

omkreddy reviewed (commented):
@steverod Thanks for the PR. Left couple of minor comments.

steverod committed:
Review feedback: Properly handle default values, assert proper values in tests.

When doing an APPEND or SUBTRACT, and the property isn't set, treat them as append/subtract on the default
value rather than "". By doing this we also don't need to worry about a null list turning into "".

steverod committed:
Review feedback: When a SUBTRACT operation empties a LIST property, remove the
property and revert to the default value.

steverod commented:
@omkreddy @rajinisivaram  thank you for the feedback! I have updated the PR to properly handle defaults and to remove the property entirely (thus reverting to default) when a SUBTRACT removes all members of the list.

omkreddy commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -598,22 +598,28 @@ class AdminManager(val config: KafkaConfig,
             throw new InvalidRequestException(s"Config value append is not allowed for config key: ${alterConfigOp.configEntry().name()}")
-          val oldValueList = configProps.getProperty(alterConfigOp.configEntry().name()).split(",").toList
+          val oldValueList = configProps.getProperty(configPropName,
+            ConfigDef.convertToString(configKeys(configPropName).defaultValue, ConfigDef.Type.LIST)).split(",").toList
```
`configKeys(configPropName).defaultValue` can return `null`. we may need to treat  `null` as empty string,

steverod replied:
Good point.

steverod committed:
Revert "Review feedback: When a SUBTRACT operation empties a LIST property, remove the"

This reverts commit 83a10ae25919e0006b1847d00ae03af8ffb2c0f4.
Stick with the behavior of a SUBTRACT of all properties resulting in an empty list property.

steverod committed:
Edge case: Handle a "null" default config value as an empty string.

steverod committed:
Simplify and Scalafy the default value handling for properties.

steverod committed:
Review feedback: Javadoc on behavior of each optype.

omkreddy commented:
add to whitelist

omkreddy reviewed (approved):
@steverod Thanks for the updates. LGTM.
